### PR TITLE
FCMS-8714 Fix delegate handling and add a bit of performance logging

### PR DESCRIPTION
**Pull request by lukas-vonblon**

### CODE REVIEW

File: src/NewRemoting/MessageHandler.cs [hunk 1]
```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
...
 			return ctor != null;
 		}
 
+		private void LogMsg(RemotingReferenceType msgType)
+		{
+			_stats[msgType]++;
+		}
+
+		/// <summary>
+		/// Write the given object to the target stream.
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:73]: Documentation is still copy-pasted

File: src/NewRemoting/MessageHandler.cs [hunk 2]
```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
...
+							}
+
+							w.Write(arg);
+						}
+					}
+				}
+				else
+				{
+					// The delegate target is a static method
+					////						w.Write(string.Empty);
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:197]: Remove commented code

File: src/NewRemoting/MessageHandler.cs [hunk 3]
```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
...
+					throw new InvalidOperationException("The delegate target is a static method");
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:198]: We should clarify the exceptions to throw here. We have a mix of `RemotingException`, `InvalidOperationException` and `NotSupportedException` for very similar errors.

File: src/NewRemoting/MessageHandler.cs [hunk 4]
```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
...
+			}
+			else if (calledMethod.IsSpecialName && calledMethod.Name.StartsWith("remove_", StringComparison.Ordinal))
+			{
+				if (del.Target != null)
+				{
+					string instanceId = _instanceManager.GetDelegateTargetIdentifier(del, remoteInstanceId);
+
+					if (_instanceManager.TryGetObjectFromId(instanceId, out var existingDelegateProxy))
+					{
+						// Create a proxy class that provides a matching event for our delegate
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:209]: Comment is incorrect (we're not creating proxies here, we're attempting to delete them)

File: src/NewRemoting/MessageHandler.cs [hunk 5]
```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
...
+					}
+
+					w.Write((int)RemotingReferenceType.RemoveEvent);
+					LogMsg(RemotingReferenceType.RemoveEvent);
+					w.Write(instanceId);
+				}
+				else
+				{
+					// The delegate target is a static method
+					throw new InvalidOperationException("The delegate target is a static method");
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:233]: The exception here is fine, but we should probably try to understand whether this is always redundant to the initial test at the beginning of the method.

File: src/NewRemoting/MessageHandler.cs [hunk 6]
```diff
@@ -101,53 +327,22 @@ public void WriteArgumentToStream(BinaryWriter w, object data, string references
...
 				w.Write((int)RemotingReferenceType.ContainerType);
+				LogMsg(RemotingReferenceType.ContainerType);
 				w.Write(data.GetType().AssemblyQualifiedName);
 				w.Write(contentType.AssemblyQualifiedName);
 				foreach (object obj in list)
 				{
 					// Recursively write the arguments
 					w.Write(true);
-					WriteArgumentToStream(w, obj, referencesWillBeSentTo);
+					WriteArgumentToStream(w, obj, referencesWillBeSentTo); // always returns true
```
Comment (pgrawehr) [inline src/NewRemoting/MessageHandler.cs:337]: Comment is no longer necessary (the method again returns void)

File: src/SampleServerClasses/MarshallableClass.cs [hunk 7]
```diff
@@ -104,7 +98,16 @@ public virtual void DoCallback()
...
+			AnEvent4?.Invoke(msg, msg, msg, Name);
+			AnEvent3?.Invoke(msg, msg, Name);
+			AnEvent2?.Invoke(msg, Name);
+			AnEvent1?.Invoke(msg);
+			AnEvent0?.Invoke();
+		}
+
+		public void DoCallbackOnEvent5(string msg)
+		{
+			AnEvent5?.Invoke(string.Empty, string.Empty, string.Empty, string.Empty, string.Empty);
```
Comment (pgrawehr) [inline src/SampleServerClasses/MarshallableClass.cs:110]: Is this intended? The parameter isn't used, which may result in the test being useless.

File: src/NewRemoting/MessageHandler.cs (hunk 1 above)
Comment (lukas-vonblon) [reply src/NewRemoting/MessageHandler.cs:73]: adapted

File: src/NewRemoting/MessageHandler.cs (hunk 3 above)
Comment (lukas-vonblon) [reply src/NewRemoting/MessageHandler.cs:198]: Clients should only catch RemotingExceptions. Unless it is a programming error we should only use this.

Comment (pgrawehr) [reply src/NewRemoting/MessageHandler.cs:196]: I agree, but then please review the exceptions in this method for this. (e.g. the use of a static method would be a programming error everywhere)

**Review by pgrawehr — DISMISSED** [review]

**Review by pgrawehr — APPROVED** [review]

**Pull request merged**